src/SqsPoller/SqsPollerHostedService.cs

```diff
@@ -62,22 +63,31 @@ private async Task Handle(string queueUrl, CancellationToken cancellationToken)
                 _logger.LogTrace(
                     "Start processing the message with id {message_id} and ReceiptHandle {receipt_handle}");
 
-                await _consumerResolver
+                await semaphore.WaitAsync(cancellationToken);
+                _consumerResolver
                     .Resolve(message, cancellationToken)
                     .ContinueWith(
-                        task => _logger.LogError(task.Exception, "Failed to handle message with id {message_id} and ReceiptHandle {receipt_handle}"),
-                        cancellationToken,
-                        TaskContinuationOptions.OnlyOnFaulted,
-                        TaskScheduler.Default)
-                    .ContinueWith(
-                        task => _logger.LogWarning("Failed to handle message with id {message_id} and ReceiptHandle {receipt_handle}. Task has been cancelled"),
-                        cancellationToken,
-                        TaskContinuationOptions.OnlyOnCanceled,
-                        TaskScheduler.Default)
-                    .ContinueWith(
-                        async _ => await DeleteMessage(queueUrl, message, cancellationToken),
+                        task =>
+                        {
+                            if (task.IsFaulted)
+                            {
+                                _logger.LogError(task.Exception,
+                                    "Failed to handle message with id {message_id} and ReceiptHandle {receipt_handle}");
+                            }
+                            else if (task.IsCanceled)
+                            {
+                                _logger.LogWarning(
+                                    "Failed to handle message with id {message_id} and ReceiptHandle {receipt_handle}. Task has been cancelled");
+                            }
+                            else
+                            {
+                                DeleteMessage(queueUrl, message);
+                            }
+
+                            semaphore.Release();
```

Comment:
Does it make sense put `semaphore.Release();` in `finally` so if `DeleteMessage` throws nothing is lost?